### ISSUE TITLE
Renamed transitive dependency breaks imports (when transitive dep is on kellnr registry)

### ISSUE THREAD

**Issue by folex** [rust-lang/cargo#11303]
### Problem

I have the following dependency tree:

1. `api_on_crates_io` – that's a lib published to crates.io, in exports a struct `Export` ([source code](https://github.com/folex/cargo-registry-violence/blob/main/api_on_crates_io/src/api.rs))
2. `user_of_that_api` – that's a lib published to [my custom registry](https://crates.fluence.dev/#/crate?name=user_of_that_api&version=0.1.0), it depends on `api_on_crates_io` like that:
```
[dependencies]
api = { package = "api_on_crates_io", version = "0.1.0" }
```
It uses `Export` from `api_on_crates_io`.

3. `user_of_user_of_api` – that's a lib which depends on `user_of_that_api` like that
```
[dependencies]
user_of_that_api = { version = "0.1.0", registry = "fluence" }
```

Now, when I try to compile `user_of_user_of_api`, compilation fails
```
folex@mac user_of_user_of_api % cargo check
    Updating `fluence` index
    Updating crates.io index
  Downloaded user_of_that_api v0.1.0 (registry `fluence`)
  Downloaded 1 crate (771 B) in 0.28s
    Checking api_on_crates_io v0.1.0
    Checking user_of_that_api v0.1.0 (registry `fluence`)
error[E0433]: failed to resolve: use of undeclared crate or module `api`
 --> /Users/folex/.cargo/registry/src/crates.fluence.dev-4cf026a00c52b3cc/user_of_that_api-0.1.0/src/lib.rs:1:5
  |
1 | use api::api::*;
  |     ^^^ use of undeclared crate or module `api`

error[E0422]: cannot find struct, variant or union type `Export` in this scope
 --> /Users/folex/.cargo/registry/src/crates.fluence.dev-4cf026a00c52b3cc/user_of_that_api-0.1.0/src/lib.rs:4:13
  |
4 |     let _ = Export {
  |             ^^^^^^ not found in this scope
  |
help: consider importing this struct
  |
1 | use api_on_crates_io::api::Export;
  |

Some errors have detailed explanations: E0422, E0433.
For more information about an error, try `rustc --explain E0422`.
error: could not compile `user_of_that_api` due to 2 previous errors
```

However, if I make `user_of_user_of_api` to depend on `user_of_that_api ` **from crates.io**, it works!
```
[dependencies]
user_of_that_api = { version = "0.1.0-cratesio" }
```
```
folex@mac user_of_user_of_api % cargo check
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on package cache
    Checking user_of_that_api v0.1.0-cratesio
    Checking user_of_user_of_api v0.1.0 (/Users/folex/Development/cargo-registry-violence/user_of_user_of_api)
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
```

### Steps

1. Check out https://github.com/folex/cargo-registry-possible-bug
2. `cd user_of_user_of_api`
3. `cargo check`

### Version

```text
% cargo version --verbose
cargo 1.65.0-nightly (4ed54cecc 2022-08-27)
release: 1.65.0-nightly
commit-hash: 4ed54cecce3ce9ab6ff058781f4c8a500ee6b8b5
commit-date: 2022-08-27
host: x86_64-apple-darwin
libgit2: 1.5.0 (sys:0.15.0 vendored)
libcurl: 7.79.1 (sys:0.4.55+curl-7.83.1 system ssl:(SecureTransport) LibreSSL/3.3.6)
os: Mac OS 12.6.0 [64-bit]
```

Same with
```text
% cargo --version --verbose
cargo 1.66.0-nightly (071eeaf21 2022-10-22)
release: 1.66.0-nightly
commit-hash: 071eeaf210708219a5a1b2c4728ca2f97df7f2ae
commit-date: 2022-10-22
host: x86_64-apple-darwin
libgit2: 1.5.0 (sys:0.15.0 vendored)
libcurl: 7.79.1 (sys:0.4.55+curl-7.83.1 system ssl:(SecureTransport) LibreSSL/3.3.6)
os: Mac OS 12.6.0 [64-bit]
```

**Comment by ehuss** [comment]
The registry index hasn't been constructed correctly.  It's important to distinguish the difference of the JSON structure passed by Cargo via the API (defined [here](https://doc.rust-lang.org/cargo/reference/registries.html#publish)) and the JSON structure as stored in the index (defined [here](https://doc.rust-lang.org/cargo/reference/registries.html#index-format)). In particular, the way renamed dependencies are handled are different (pay close attention to the "name" field and the "package" field).

The fluence.dev index has an entry that looks like this:

```javascript
{
    "name": "user_of_that_api",
    "vers": "0.1.0",
    "cksum": "903975f7eca33517910b73a7c2b010c416f1da7b02942b584dcfc3f16407f218",
    "deps":
    [
        {
            "name": "api_on_crates_io",
            "req": "^0.1.0",
            "features":
            [],
            "optional": false,
            "default_features": true,
            "target": null,
            "kind": "normal",
            "registry": "https://github.com/rust-lang/crates.io-index",
            "explicit_name_in_toml": "api"
        }
    ],
    "features":
    {},
    "authors":
    [],
    "description": "testing how package renaming works across different registries",
    "documentation": null,
    "homepage": null,
    "readme": null,
    "readme_file": null,
    "keywords":
    [],
    "categories":
    [],
    "license": "MIT",
    "license_file": null,
    "repository": null,
    "badges":
    {},
    "yanked": false,
    "links": null
}
```

Notice that the `deps` have a "explicit_name_in_toml" field, which is not part of the index schema.

**Comment by secana** [comment]
The issue is fixed in Kellnr 2.1.2. Thanks for reporting @ehuss and thanks for pointing out where the problem is @folex. I suggest to close the issue here, as it was not a Cargo issue but one in Kellnr.

**Comment by folex** [comment]
@secana thank you!
